### PR TITLE
More work on FAT / System.IO

### PR DESCRIPTION
**Pull request by fanoI**
The objective of this PR is to leverage the fact that in Net Core a lot of code that we needed to plug for Net Framework is not needed anymore there is Managed Code or a low level class called "FileSystem" in which all platform dependent code is concentrated.

- [x]  Introduces new plug CosmosFileSystem that plugs any instance of the FileSystem class
- [x]  Add DirectoryInfo support
- [x]  Remove unneeded plugs from the Directory class (it should be all managed code calling FileSystem class)
- [x]  Add FileInfo support
- [x] Remove unneeded plugs from File class (it should be all managed code calling StreamWriter / StreamReader classes)
- [x] Plug StreamWriter
- [x] Plug StreamReader
  
Blocked by https://github.com/CosmosOS/IL2CPU/pull/32

### CODE REVIEW

File: source/Cosmos.System2_Plugs/System/Runtime/CompilerServicesImpl.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using Cosmos.IL2CPU.API.Attribs;
+
+namespace Cosmos.System_Plugs.System.Runtime
+{
+    [Plug(Target = typeof(global::System.Runtime.CompilerServices.RuntimeHelpers))]
+    public static class RuntimeHelpersImpl
+    {
+        /* For RAT in Cosmos the size of Reference Types is 8 on x86 / 32 bit too...
+         * String object header has two reference types so the data part is at offset 16
```
Comment (jp2masa) [inline source/Cosmos.System2_Plugs/System/Runtime/CompilerServicesImpl.cs]: This comment is not totally correct.
1. References are: `(currentAddress << sizeof(IntPtr)) | nextAddress`, where `currentAddress` and `nextAddress` are `IntPtr`.
2. The data part starts at offset 12, but the first char is at offset 16, which is what this property returns. From the comment in CoreCLR (notice that it skips over the string length, which is the first 4 bytes after the object header):
>// Number of bytes from the address pointed to by a reference to
// a String to the first 16-bit character in the String.  Skip 
// over the MethodTable pointer, & String 
// length.

**Comment by geomtech** [comment]
You said that DirectoryInfo will fixe https://github.com/CosmosOS/Cosmos/issues/747, so it will be fixed with this pull request ?

**Comment by jp2masa** [comment]
This pull request improves the support of Cosmos for file system methods in the BCL. #747 is a FAT bug.

**Comment by fanoI** [comment]
If I remember well "folder.system" works correctly in this PR but not ".folder".

**Comment by jp2masa** [comment]
That's because the current FAT implementation considers entries that start with a dot as dot entries. I'm not sure about the correct behaviour, but that needs to be reviewed.

**Comment by fanoI** [comment]
It is in any case not object of this PR that is more use more of the Net Core managed code possible.

File: source/Cosmos.System2_Plugs/System/IO/PathImpl.cs [hunk 2]
```diff
@@ -242,7 +242,7 @@ public static string GetTempFileName()
 
         public static string GetTempPath()
         {
-            return @"\Temp";
+            return @"0:\Temp";
```
Comment (jp2masa) [inline source/Cosmos.System2_Plugs/System/IO/PathImpl.cs]: Why is this change needed?

File: source/Cosmos.System2_Plugs/System/IO/FileInfoImpl.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+using IL2CPU.API.Attribs;
+using Cosmos.System;
+using Cosmos.System.FileSystem.VFS;
+
+namespace Cosmos.System2_Plugs.System.IO
+{
+    [Plug(Target = typeof(FileInfo))]
+    public static class FileInfoImpl
+    {
+        /* Optimize this: CosmosVFS should expose an attribute without the need to open the file for reading... */
```
Comment (jp2masa) [inline source/Cosmos.System2_Plugs/System/IO/FileInfoImpl.cs]: I think that the FAT driver has support for entry sizes. The VFS has to provide a way of getting the size, but it doesn't need to be done in this PR.

File: source/Cosmos.System2_Plugs/System/IO/FileImpl.cs [hunk 4]
```diff
@@ -16,205 +16,32 @@ namespace Cosmos.System_Plugs.System.IO
...
-            }
+            Global.mFileSystemDebugger.SendInternal("Writing contents");
 
-            var xEntry = VFSManager.CreateFile(aFile);
-            if (xEntry == null)
+            using (var xSW = new StreamWriter(aFile))
             {
-                return null;
+                foreach (var current in contents)
+                    xSW.WriteLine(current);
```
Comment (jp2masa) [inline source/Cosmos.System2_Plugs/System/IO/FileImpl.cs]: Nit: please use braces.

File: source/Cosmos.Core_Plugs/Cosmos.Core_Plugs.csproj [hunk 5]
```diff
@@ -22,4 +22,8 @@
     <Folder Include="Properties\" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Include="System\Threading\TaskImpl.cs" />
```
Comment (jp2masa) [inline source/Cosmos.Core_Plugs/Cosmos.Core_Plugs.csproj]: Why is it included as None?

File: Tests/Cosmos.Compiler.Tests.Bcl/Kernel.cs [hunk 6]
```diff
@@ -42,10 +42,11 @@ protected override void Run()
                 BitConverterTest.Execute();
                 UnsafeCodeTest.Execute();
                 DelegatesTest.Execute();
+                EncodingTest.Execute();
                 RandomTests.Execute();
 
-                System.Collections.Generic.ListTest.Execute();
-                System.Collections.Generic.QueueTest.Execute();
+                //System.Collections.Generic.ListTest.Execute();
+                //System.Collections.Generic.QueueTest.Execute();
```
Comment (jp2masa) [inline Tests/Cosmos.Compiler.Tests.Bcl/Kernel.cs]: Why are these tests commented?

File: source/Cosmos.System2_Plugs/System/IO/PathImpl.cs (hunk 2 above)
Comment (fanoI) [reply source/Cosmos.System2_Plugs/System/IO/PathImpl.cs]: Testing on Windows the value returned is indeed "C:/temp", it made some low level thing not work... Path or Directory I cannot remember correctly.

File: Tests/Cosmos.Compiler.Tests.Bcl/Kernel.cs (hunk 6 above)
Comment (fanoI) [reply Tests/Cosmos.Compiler.Tests.Bcl/Kernel.cs]: I've simply forgotten to re-enable them, I do this now.

File: source/Cosmos.Core_Plugs/Cosmos.Core_Plugs.csproj (hunk 5 above)
Comment (fanoI) [reply source/Cosmos.Core_Plugs/Cosmos.Core_Plugs.csproj]: I think this has been done automatically by Visual Studio, where should it be?

File: source/Cosmos.System2_Plugs/System/IO/FileInfoImpl.cs (hunk 3 above)
Comment (fanoI) [reply source/Cosmos.System2_Plugs/System/IO/FileInfoImpl.cs]: Yes I think so, this can be a "starting point".

File: source/Cosmos.System2_Plugs/System/IO/PathImpl.cs (hunk 2 above)
Comment (jp2masa) [reply source/Cosmos.System2_Plugs/System/IO/PathImpl.cs]: The root drive should represent the system drive I think. "0:\\" may not be the root drive, so this needs to be fixed in the future.

File: source/Cosmos.Core_Plugs/Cosmos.Core_Plugs.csproj (hunk 5 above)
Comment (jp2masa) [reply source/Cosmos.Core_Plugs/Cosmos.Core_Plugs.csproj]: It shouldn't be added at all.
